### ISSUE TITLE
S3217 should not suggest to make string an object in foreach construct

### ISSUE THREAD

**Issue by jessehouwing** [SonarSource-VisualStudio/sonarlint-roslyn-analyzer#212]
When using foreach on an "old style" IEnumerable, it's not useful to suggest to make the iterator variable of type object.

Severity	Code	Description	Project	File	Line	Category
Warning	S3217	Make "s" a "object".	Aggregator.Core	C:\Users\jhouw\Source\Repos\TfsAggregator\tfsaggregator\Aggregator.Core\Extensions\WorkItemExtensions.cs	18	SonarQube

     foreach (string s in wi.Validate())
     {}

I can currently bypass this warning by adding `.Cast<string>()` like so:

     foreach (string s in wi.Validate().Cast<string>())
     {}

But that adds nothing, as this is what foreach already does under the hood.

**Comment by dbolkensteyn** [comment]
@jessehouwing you mean that `Validate()` is returning for example an `ArrayList` which is not generic?

It seems to me like a valuable diagnostics, but the fix you'll want to apply is not to add a `.Cast<string>()` indeed - which doesn't bring much value - but to go and update `.Validate()` itself to return for example a `IEnumerable<string>`.

Do you think that we should simply improve the issue message to suggest the right action?

It indeed doesn't make a lot of sense to make `s` of type `object`.

**Comment by njannink** [comment]
For these old-style enumerables it might be good to do a `OfType<string>()` to make sure your code doesn't break when somebody desides to return something else in the `Validate()` method except of a string.

**Comment by jessehouwing** [comment]
Validate is a method from the TFS Client Object model and not in my circle of control to update. If the code were in my circiel of control I'd be happy to see a warning which instructs me to change the returntype of the method I'm calling. But as it stands this warning is useless. What do you want me to do after changing it to "object"?

      foreach (object o in wi.Validate())
      {
          string s = (string)o; // results in the exact same IL as using string s in the foreach.
          
          // or something like this? And what if it's null? Throw an exception? Ignore this edge case?
          string s = object as string;
          if (s != null)
      }

@njannink I would use `.Cast` instead of `.OfType`, as `.OfType` could accidentally cause objects to be missed or filtered out without any future indication. I prefer the code to break when the method suddenly returns something different. That's better than for my code to suddenly change its perceived behavior.

The call to .Cast is not required, as the implementation of `foreach` in C# already does the cast and already throws the InvalidCastexception. Any other code is more complex and doesn't improve the readability nor the stability of my code.

**Comment by dbolkensteyn** [comment]
Looks like this rule, as many others, is a bit controversial :)

It might also be an option to check whether or not `wi.Validate()` is defined within the same project / solution - and only report if we're sure that you have the ability to edit it. This however won't handle company libraries for instance on which you've added a NuGet dependency - you still could get that code fixed.

**Comment by jessehouwing** [comment]
I'd expect a rule to trigger on any property or field that returns an old style Collection, be it ArrayList or any other "untyped" collection. But not on the usage of such a property in the foreach.

**Comment by tvsonar** [comment]
I'm not convinced about this approach. This rule is about the unexpected behavior of `foreach`, and not about restricting collection types to generic collections. (I could imaging such a rule with info level severity.)

The problem with `foreach` is that it might not behave as you'd expect: 
> the real feature is “assert that every item in the sequence is of type V and crash if it is not”, whereas 
> I believed it was “for every element in this sequence of type V…”.

[from Eric Lipper](http://ericlippert.com/2013/07/22/why-does-a-foreach-loop-silently-insert-an-explicit-conversion/). 

I think there can be two fixes to this issue:
* add `.OfType<T>()`,
* add `.Cast<T>()`.
The first one changes the functionality of the code, the second one doesn't. Both seems like a valid choice. If you had the feeling that `foreach` was filtering for you, then use `.OfType<T>()`. If you knew that it was not filtering add the `.Cast<T>()` to explicitly show the assumption that the collection should only contain elements of type `T`. 

But what's sure is that we need to improve the issue message.

**Comment by jessehouwing** [comment]
> *Side note*: I'm amazed that Eric Lipper made that assumption.

To me `foreach` behaves exactly as I expect it. It has been so since C#1.  Plus I haven't seen any developer fall into this trap in years.

I do not like the `.Cast<T>` on it, as it would add performance overhead of the call and the additional cast. 

Plus, it's very evident when it happens. It adds noise to the static analysis with something the code already does and does well. It introduces an implicit cast to  the type specified in the `foreach` construct.

Also note that the extension methods are not available in all versions of the .NET framework. So if this is going to be the suggestion, make sure that a test is done for the availability of these methods.

My suggestion would be to provide an informational hint, not a warning.

**Comment by jessehouwing** [comment]
As I suspected `foreach` will explicitly convert in the below test suite. interestingly enough an implicit conversion fails. `Cast` and `OfType` both fail in this case.

So introducing this change will be breaking as well as not supported pre .NET 3.5 and in Silverlight.
```
using System;
using System.Linq;

using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace UnitTestProject2
{
    [TestClass]
    public class UnitTest1
    {
        [TestMethod]
        public void ForeachWillExplicitlyConvertBtoA()
        {
            B[] bs = new[] { new B(), new B() };

            bool enteredLoop = false;
            foreach (A a in bs)
            {
                enteredLoop = true;
                Assert.AreEqual(a.Text, "explicit from B");
            }
            Assert.IsTrue(enteredLoop, "Did not enter the loop as expected");
        }

        [TestMethod]
        public void ForeachCastWillExplicitlyConvertBtoA()
        {
            B[] bs = new[] { new B(), new B() };

            bool enteredLoop = false;
            foreach (A a in bs.Cast<A>())
            {
                enteredLoop = true;
                Assert.AreEqual(a.Text, "explicit from B");
            }
            Assert.IsTrue(enteredLoop, "Did not enter the loop as expected");
        }

        [TestMethod]
        public void ForeachOfTypeWillExplicitlyConvertBtoA()
        {
            B[] bs = new[] { new B(), new B() };

            bool enteredLoop = false;
            foreach (A a in bs.OfType<A>())
            {
                enteredLoop = true;
                Assert.AreEqual(a.Text, "explicit from B");
            }

            Assert.IsTrue(enteredLoop, "Did not enter the loop as expected");
        }





        [TestMethod]
        public void ForeachWillImplicitlyConvertAtoB()
        {
            A[] @as = new[] { new A(), new A() };

            bool enteredLoop = false;
            foreach (B b in @as)
            {
                enteredLoop = true;
                Assert.AreEqual(b.Text, "implicit from A");
            }
            Assert.IsTrue(enteredLoop, "Did not enter the loop as expected");
        }

        [TestMethod]
        public void ForeachCastWillExplicitlyConvertAtoB()
        {
            A[] @as = new[] { new A(), new A() };

            bool enteredLoop = false;
            foreach (B b in @as.Cast<B>())
            {
                enteredLoop = true;
                Assert.AreEqual(b.Text, "implicit from A");
            }
            Assert.IsTrue(enteredLoop, "Did not enter the loop as expected");
        }

        [TestMethod]
        public void ForeachOfTypeWillExplicitlyConvertAtoB()
        {
            A[] @as = new[] { new A(), new A() };

            bool enteredLoop = false;
            foreach (B b in @as.OfType<B>())
            {
                enteredLoop = true;
                Assert.AreEqual(b.Text, "implicit from A");
            }

            Assert.IsTrue(enteredLoop, "Did not enter the loop as expected");
        }
    }

    public class A
    {
        public string Text { get; set; }
        public A()
        {
        }

        public A(string text)
        {
            Text = text;
        }

        public static implicit operator B(A a)
        {
            return new B();
        }
    }

    public class B
    {
        public string Text { get; set; }

        public B()
        {
        }

        public B(string text)
        {
            Text = text;
        }

        public static explicit operator A(B b)
        {
            return new A("explicit from B");
        }
    }


}
```

**Comment by jessehouwing** [comment]
Ahh, `Cast` uses the `as Type` operator, which indeed fails as it won't perform the conversion, only `(Type)` notation will perform the conversion.

**Comment by tvsonar** [comment]
Thanks, for the thorough analysis. Yep, it's not that surprising that the casts are not working the same way in the extensions as in the `foreach`. And this suggests me that we should only propose the `.OfType<T>()` codefix, which is changing the behavior as expected.

**Comment by jessehouwing** [comment]
In that case make sure it's not a warning. This will trigger like crazy on any codebase out there and the change in behavior may not be well understood or intended. 

I'm still very hesitant to suggest this to users, as it would not surface clearly in production if the type of the enumerable has changed. The method will fail silently, which can be far more problematic than a hard-stop.

Consider a class that fetches a set of rules to apply on  a payment, if all rules approve, the money can be transferred:
      
      bool approve = true;
      foreach (Rule rule in approval.Rules)
      {
            approve &= rule.Approve(amount);
      }

Now if the type of the bad, bad, bad ArrayList Rules changes to something else, all payments will automatically be approved. If the suggestion would be applied.

It would crash (or at least fail on this specific method) if `foreach` or `.Cast` is used.

This is why I'd be **very** careful suggesting users to hit <kbd>{ctrl}</kbd>-<kbd>.</kbd>, <kbd>{enter}</kbd>.

**Comment by tvsonar** [comment]
Do you think there is a better fix? If the fix is contraversial, then we might not even implement it.

I think this produces equivalent code to the foreach cast: `foreach(string item in arraylist.Select(item=>(string)item))`, but I don't think this should be suggested as the fixed code, because it is a lot less clear then the `.Cast<T>()`, which we already excluded. (Although I don't think it is a very common scenario to have implicit/explicit casts.)

I don't think the severity should be reduced, because this rule can catch possible bugs. It should draw attention to the possibly wrong code.
